waspc/src/Wasp/AppSpec/Valid.hs

```diff
@@ -0,0 +1,91 @@
+{-# LANGUAGE TypeApplications #-}
+
+module Wasp.AppSpec.Valid
+  ( validateAppSpec,
+    ValidationError (..),
+    getApp,
+    isAuthEnabled,
+  )
+where
+
+import Data.List (find)
+import Data.Maybe (isJust)
+import Wasp.AppSpec (AppSpec)
+import qualified Wasp.AppSpec as AS
+import Wasp.AppSpec.App (App)
+import qualified Wasp.AppSpec.App as App
+import qualified Wasp.AppSpec.App.Auth as Auth
+import Wasp.AppSpec.Core.Decl (takeDecls)
+import qualified Wasp.AppSpec.Entity as Entity
+import qualified Wasp.AppSpec.Entity.Field as Entity.Field
+import qualified Wasp.AppSpec.Page as Page
+
+data ValidationError = GenericValidationError String
+  deriving (Show, Eq)
+
+validateAppSpec :: AppSpec -> [ValidationError]
+validateAppSpec spec =
+  case validateExactlyOneAppExists spec of
+    Just err -> [err]
+    Nothing ->
+      -- NOTE: We check these only if App exists because they all rely on it existing.
+      concat
+        [ validateAppAuthIsSetIfAnyPageRequiresAuth spec,
+          validateAuthUserEntityHasCorrectFields spec
+        ]
+
+validateExactlyOneAppExists :: AppSpec -> Maybe ValidationError
+validateExactlyOneAppExists spec =
+  case AS.takeDecls @App (AS.decls spec) of
+    [_] -> Nothing
+    apps ->
+      Just $
+        GenericValidationError $
+          "Expected exactly 1 'app' declaration, but found " ++ show (length apps) ++ "."
+
+validateAppAuthIsSetIfAnyPageRequiresAuth :: AppSpec -> [ValidationError]
+validateAppAuthIsSetIfAnyPageRequiresAuth spec =
+  if anyPageRequiresAuth && not (isAuthEnabled spec)
+    then
+      [ GenericValidationError
+          "Expected app.auth to be defined since there are Pages with authRequired set to true."
+      ]
+    else []
+  where
+    anyPageRequiresAuth = any (isJust . Page.authRequired) (snd <$> AS.getPages spec)
+
+validateAuthUserEntityHasCorrectFields :: AppSpec -> [ValidationError]
+validateAuthUserEntityHasCorrectFields spec = case App.auth (snd $ getApp spec) of
+  Nothing -> []
+  Just auth ->
+    let userEntity = snd $ AS.resolveRef spec (Auth.userEntity auth)
+        fields = Entity.getFields userEntity
+        maybeEmailField = find ((== "email") . Entity.Field.fieldName) fields
+        maybePasswordField = find ((== "password") . Entity.Field.fieldName) fields
+     in concat
+          [ case maybeEmailField of
+              Just emailField
+                | Entity.Field.fieldType emailField == Entity.Field.FieldTypeScalar Entity.Field.String -> []
```

Comment:
```suggestion
                | Entity.Field.fieldType emailField == (Entity.Field.FieldTypeScalar Entity.Field.String) -> []
```
is it being evaluated like this?

Reply:
Yes! What I learned is that binary operators normally have low precedence, so you mostly don't need to group whatever is left or right of them. Especially not application, application has very high precedence.